File: src/main/java/com/appdetex/sampleparserjavaproject/utils/GoogleAppStoreScraper.java
package com.appdetex.sampleparserjavaproject.utils;

import java.io.IOException;
import java.text.ParseException;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import com.appdetex.sampleparserjavaproject.domain.ItemProperties;


public class GoogleAppStoreScraper implements Scraper {
	private static String title;
	private static String description;
	private static String publisher;
	private static String price;
	private static Double rating;

	public ItemProperties scrapeUrl(String url)  throws ParseException, IOException {
        Document doc = Jsoup.connect(url).get();
        return scrapeDocument(doc);
	}

	public ItemProperties scrapeDocument(Document doc) throws ParseException {

		Elements itemProps = doc.getElementsByAttribute("ItemProp");

        for (Element prop : itemProps) {
        	if(prop.attr("ItemProp").equalsIgnoreCase("price")) {
        		price = prop.attr("content").trim();
        	} else if(prop.attr("ItemProp").equalsIgnoreCase("ratingValue")) {
        		rating = Double.parseDouble(prop.attr("content").trim());
        	} else if(prop.attr("ItemProp").equalsIgnoreCase("name") ) {
        		if(prop.hasClass("document-title")) {
        			title = prop.text().trim();
        		} else if(!prop.attr("content").equalsIgnoreCase("Android")){
        			publisher = prop.text().trim();
        		}
        	} else if(prop.attr("ItemProp").equalsIgnoreCase("description") ) {
      			description = prop.select("[jsname]").first().childNodes().get(0).toString().trim();
           	}
        }
        return new ItemProperties(title, publisher, description, price, rating);
	}
}

File: src/main/java/com/appdetex/sampleparserjavaproject/GoogleAppStoreScraper.kt
package com.appdetex.sampleparserjavaproject

import org.apache.commons.validator.routines.UrlValidator
import org.jsoup.Jsoup


/**
 * Fetch, Scrape, and return a data object of the request application
 * from the Google App Store.
 *
 * Validates the passed url is a vaild url.
 * Attempts to scrape page assuming the format has not changed since 2019-04-20 @ 1802MST
 *
 * @param url Full url to scrape from Google Store
 *
 */
class GoogleAppStoreScraper(url: String){
    private val target = url


    /**
     * Scrapes App page
     * @return GoogleAppStoreInfo either with default values or Results of page scrape
     */
    private fun scrapedApp(): GoogleAppStoreInfo {
        var appInfo = GoogleAppStoreInfo()

        Jsoup.connect(target).ignoreHttpErrors(true).get().run {
            select("[class=\"LXrl4c\"]").forEachIndexed { _, element ->
                val name = element.select("[itemprop=\"name\"]").text()
                val fullDesc = element.select("[itemprop=\"description\"]").text()
                val description = fullDesc.split('\n').first()

                val publisher = element.select("div:contains(offered by) > span > div > span").text()


                val price = element.select("[itemprop=\"price\"]").attr("content")
                val rating = element.select("[aria-label*=stars out of five stars]").text().toFloat()

                appInfo = GoogleAppStoreInfo(
                        appTitle = name,
                        price = price,
                        rating = rating,
                        publisher = publisher,
                        description = description
                )

            }
        }

        return appInfo
    }

    /**
     * Validates that URL is a "valid" url.
     * @return boolean
     */
    private fun validUrl(): Boolean {
        val urlValidator = UrlValidator()
        return urlValidator.isValid(this.target)
    }


    /**
     * If valid url, scrapes store
     * Otherwise, returns default values from GoogleApp
     * @return GoogleAppStoreInfo
     */
    fun result(): GoogleAppStoreInfo {
        if( validUrl() ) {
            return scrapedApp()
        }

        return GoogleAppStoreInfo()
    }
}
